File: apps/desktop/src/modules/db/filters.ts
/** The parts of a filter bar that don't care which database is underneath: how many values an
 * operator reads out of one text box, how that box is split, and what a row of the bar holds
 * while it is being edited. The operator sets themselves are per-database — see
 * `src/mysql/filters.ts` and `src/mongo/filters.ts`. */

/** How many values an operator reads out of the one text box the filter row gives it. */
export type FilterArity =
  /** None — `IS NULL` and its kind stand on their own, and the box is left disabled. */
  | "none"
  /** One, taken as typed. */
  | "one"
  /** A comma-separated list, for `IN` / `NOT IN`. */
  | "list"
  /** Two bounds, `min,max`, for `BETWEEN` / `NOT BETWEEN`. */
  | "pair";

/** One entry of a database's operator list: the id the backend matches on, and how its value is
 * read. The label lives in the translations, keyed by the same id. */
export interface FilterOperatorSpec {
  id: string;
  arity: FilterArity;
}

/** One row of the bar while it is being edited — a condition plus the two things only the UI
 * cares about: whether it is switched on, and what tells it apart from an identical row. */
export interface FilterRow<Op extends string = string> {
  /** Identity for React and for the edit handlers, since two rows may otherwise be equal and a
   * row's position changes as the ones above it are removed. */
  id: number;
  /** The checkbox at the head of the row: off leaves the condition written down but unapplied. */
  enabled: boolean;
  column: string;
  operator: Op;
  value: string;
}

/** One condition sent to the server. They are ANDed together; the value is always text, and the
 * operator is what says how to read it. */
export interface QueryFilter<Op extends string = string> {
  column: string;
  operator: Op;
  value: string;
}

let nextRowId = 1;

/** Looks an operator's arity up in a database's own list. Built once per list rather than
 * searched each time, since every render of the bar asks for every row's operator. */
export function arityLookup<Op extends string>(
  operators: readonly FilterOperatorSpec[],
): (operator: Op) => FilterArity {
  const table = new Map(operators.map((op) => [op.id, op.arity]));
  return (operator) => table.get(operator) ?? "one";
}

/**
 * Splits an `IN`/`BETWEEN` value into its items — `1,2,3` or `abc,xyz`, each item trimmed.
 *
 * An item may be wrapped in single or double quotes, which is how a value that itself holds a
 * comma (or spaces that matter) gets through: the quotes come off and what was inside them is
 * taken as-is. Empty items are dropped, so a trailing comma costs nothing — but a quoted `''`
 * stays, being the only way to put an empty string in a list.
 *
 * This mirrors `split_list` in `src-tauri/src/db/filters.rs`, which is what actually builds the
 * conditions; the two have to agree on how many items a value holds or a row would be sent as
 * complete and come back matching something else.
 */
export function splitFilterList(raw: string): string[] {
  const items: string[] = [];
  let current = "";
  let quote: string | null = null;
  let quoted = false;

  function flush() {
    const item = quoted ? current : current.trim();
    if (quoted || item !== "") items.push(item);
    current = "";
    quoted = false;
  }

  for (const ch of raw) {
    if (quote !== null) {
      if (ch === quote) quote = null;
      else current += ch;
      continue;
    }
    if (ch === "'" || ch === '"') {
      // Whitespace between the comma and the opening quote is padding around the item rather than
      // part of it — a quoted item is not trimmed afterwards, so `'a', 'b'` would otherwise ask
      // for a value beginning with a space.
      if (current.trim() === "") current = "";
      quote = ch;
      quoted = true;
      continue;
    }
    if (ch === ",") {
      flush();
      continue;
    }
    // The same padding on the other side of the closing quote.
    if (quoted && ch.trim() === "") continue;
    current += ch;
  }
  flush();
  return items;
}

/** Whether a row has everything its operator needs to be worth sending. A row that hasn't been
 * filled in yet is left out of the query entirely rather than matched literally — the bar opens
 * with an empty `id =` row, and that must not mean "the rows whose id is the empty string". */
export function isFilterComplete(arity: FilterArity, value: string): boolean {
  switch (arity) {
    case "none":
      return true;
    case "list":
      return splitFilterList(value).length > 0;
    case "pair":
      return splitFilterList(value).length >= 2;
    default:
      return value !== "";
  }
}

/** The column a row starts on: the id column when there is one — it is what a lookup is nearly
 * always by, and Mongo's `_id` counts — and otherwise the first column, so the row is never left
 * pointing at nothing. */
function startingColumn(columns: string[]): string {
  return columns.find((c) => c.toLowerCase() === "id" || c.toLowerCase() === "_id") ?? columns[0] ?? "";
}

/** One row of the bar for a condition that is already known — following a foreign key writes the
 * whole of it, column, operator and value together, rather than leaving anything to be typed. */
export function filterRowFor<Op extends string>(column: string, operator: Op, value: string): FilterRow<Op> {
  return { id: nextRowId++, enabled: true, column, operator, value };
}

export function createFilterRow<Op extends string>(columns: string[], operator: Op): FilterRow<Op> {
  return filterRowFor(startingColumn(columns), operator, "");
}

/** What the bar holds when a table is first opened: an empty `id =` row, ready for the lookup
 * that is about to be typed into it. A table with no id column starts with no rows at all —
 * there is no column to guess at, and an arbitrary one would only be in the way. */
export function initialFilterRows<Op extends string>(columns: string[], operator: Op): FilterRow<Op>[] {
  const hasId = columns.some((c) => c.toLowerCase() === "id" || c.toLowerCase() === "_id");
  return hasId ? [createFilterRow(columns, operator)] : [];
}

/** The conditions that actually reach the query: the rows that are switched on and filled in.
 * A row whose operator still wants a value is dropped rather than sent — see
 * {@link isFilterComplete}. */
export function toQueryFilters<Op extends string>(
  rows: FilterRow<Op>[],
  arityOf: (operator: Op) => FilterArity,
): QueryFilter<Op>[] {
  return rows
    .filter((row) => row.enabled && row.column !== "" && isFilterComplete(arityOf(row.operator), row.value))
    .map(({ column, operator, value }) => ({ column, operator, value }));
}

File: apps/desktop/src-tauri/src/modules/db/drivers/filters.rs
//! The part of the grid's filter bar that reads the same whichever database is underneath: how
//! the one text box a row gives an `IN`/`BETWEEN` operator is split into its items. Turning those
//! items into a query is per-database — `build_where` in `mysql.rs`, `build_filter` in `mongo.rs`.
//! What both SQL servers spell the same way is here too.

/// One item of a split value, and whether it arrived in quotes. The quotes are what a caller that
/// infers types from the text goes by: `123` is a number, `'123'` is that number spelled out.
pub struct ListItem {
    pub text: String,
    pub quoted: bool,
}

/// Splits an `IN`/`BETWEEN` value into its items: comma-separated, with each item trimmed.
///
/// An item may be wrapped in single or double quotes, which is how a value that itself contains a
/// comma (or leading/trailing spaces that matter) gets through: the quotes are stripped and the
/// text inside them is taken as-is. Empty items are dropped, so `1,2,` is two items — but a
/// quoted `''` is kept, as that is the only way to ask for an empty string in a list.
///
/// The frontend mirrors this in `src/filters.ts` to decide whether a row has enough to be worth
/// sending; the two must agree on how many items a value holds.
pub fn split_list_parts(raw: &str) -> Vec<ListItem> {
    let mut items: Vec<ListItem> = Vec::new();
    let mut current = String::new();
    let mut quote: Option<char> = None;
    let mut quoted = false;

    let mut flush = |current: &mut String, quoted: &mut bool| {
        let text = if *quoted {
            std::mem::take(current)
        } else {
            current.trim().to_string()
        };
        if *quoted || !text.is_empty() {
            items.push(ListItem {
                text,
                quoted: *quoted,
            });
        }
        current.clear();
        *quoted = false;
    };

    for ch in raw.chars() {
        match quote {
            Some(q) if ch == q => quote = None,
            Some(_) => current.push(ch),
            None if ch == '\'' || ch == '"' => {
                // Whitespace between the comma and the opening quote is padding around the item
                // rather than part of it — a quoted item is not trimmed afterwards, so `'a', 'b'`
                // would otherwise ask for a value beginning with a space.
                if current.trim().is_empty() {
                    current.clear();
                }
                quote = Some(ch);
                quoted = true;
            }
            None if ch == ',' => flush(&mut current, &mut quoted),
            // The same padding on the other side of the closing quote.
            None if quoted && ch.is_whitespace() => {}
            None => current.push(ch),
        }
    }
    flush(&mut current, &mut quoted);
    items
}

/// {@link split_list_parts} for a caller that binds every item as text anyway and so has no use
/// for knowing which of them were quoted.
pub fn split_list(raw: &str) -> Vec<String> {
    split_list_parts(raw)
        .into_iter()
        .map(|item| item.text)
        .collect()
}

/// Takes the quotes off a value that is wrapped in a matching pair of them, which is how a single
/// value is spelled when it is to be read as text and nothing else. Anything not so wrapped comes
/// back as `None` and is left to whatever the caller infers from it.
pub fn unquote(raw: &str) -> Option<&str> {
    let mut chars = raw.chars();
    let first = chars.next()?;
    if first != '\'' && first != '"' {
        return None;
    }
    let last = chars.next_back()?;
    if last != first {
        return None;
    }
    Some(&raw[first.len_utf8()..raw.len() - last.len_utf8()])
}

/// The wildcards out of a value going into a `LIKE`.
///
/// MySQL and PostgreSQL both take `\` as the escape character in a `LIKE` pattern unless told
/// otherwise, so this is one function rather than two — the backslash first, or it would escape
/// the escapes added after it.
///
/// A `contains` filter for `50%` must match the text `50%`, not "starts with 50". The user is
/// filtering a column, not writing a pattern; the bar has no syntax for one and this is what keeps
/// it that way.
pub fn escape_like(value: &str) -> String {
    value
        .replace('\\', "\\\\")
        .replace('%', "\\%")
        .replace('_', "\\_")
}

#[cfg(test)]
mod tests {
    use super::{escape_like, split_list_parts, unquote};

    fn texts(raw: &str) -> Vec<String> {
        split_list_parts(raw)
            .into_iter()
            .map(|item| item.text)
            .collect()
    }

    #[test]
    fn splits_on_commas_and_trims_what_is_left() {
        assert_eq!(texts("1, 2 ,3"), ["1", "2", "3"]);
        // A trailing comma names no item, so `1,2,` is two items and not three.
        assert_eq!(texts("1,2,"), ["1", "2"]);
        assert_eq!(texts("   "), Vec::<String>::new());
    }

    /// Quotes are how a value carrying a comma — or spaces that matter — gets through in one
    /// piece, and the only way to ask for the empty string.
    #[test]
    fn quotes_hold_a_value_together() {
        assert_eq!(texts("'a,b', c"), ["a,b", "c"]);
        assert_eq!(texts("' padded '"), [" padded "]);
        assert_eq!(texts("''"), [""]);
        assert_eq!(texts(r#""double", 'single'"#), ["double", "single"]);
    }

    /// The spaces a list is typed with are around the items, not in them: only what stands between
    /// the quotes is the value.
    #[test]
    fn ignores_the_spacing_around_a_quoted_item() {
        assert_eq!(texts("'a' , 'b'"), ["a", "b"]);
        assert_eq!(texts("  'a'  "), ["a"]);
    }

    /// Whether an item arrived quoted is what a caller inferring types goes by: `123` is a number,
    /// `'123'` is that number spelled out.
    #[test]
    fn reports_which_items_were_quoted() {
        let parts = split_list_parts("123, '123'");
        assert_eq!(parts.len(), 2);
        assert!(!parts[0].quoted);
        assert!(parts[1].quoted);
    }

    #[test]
    fn unquotes_only_a_matching_pair() {
        assert_eq!(unquote("'x'"), Some("x"));
        assert_eq!(unquote("\"x\""), Some("x"));
        assert_eq!(unquote("''"), Some(""));
        assert_eq!(unquote("'x\""), None);
        assert_eq!(unquote("x"), None);
        assert_eq!(unquote("'"), None);
    }

    #[test]
    fn like_wildcards_are_escaped_so_a_filter_is_not_a_pattern() {
        // A `contains` filter for `50%` means the text `50%`, not "starts with 50".
        assert_eq!(escape_like("50%"), r"50\%");
        assert_eq!(escape_like("a_b"), r"a\_b");
    }

    #[test]
    fn the_escape_character_is_escaped_first() {
        // Backslash last would escape the escapes already added, and `\%` would come out as a
        // literal backslash followed by a live wildcard.
        assert_eq!(escape_like(r"a\b"), r"a\\b");
        assert_eq!(escape_like(r"\%"), r"\\\%");
    }

    #[test]
    fn text_with_no_wildcards_in_it_is_left_alone() {
        assert_eq!(escape_like("plain text"), "plain text");
    }
}
